src/main/java/com/mongodb/hibernate/internal/translate/AbstractMqlTranslator.java

```diff
@@ -580,17 +577,93 @@ public void visitTuple(SqlTuple sqlTuple) {
+            fieldUpdates.add(new AstFieldUpdate(fieldPath, fieldValue));
+        }
+        astVisitorValueHolder.yield(
+                COLLECTION_MUTATION,
+                new AstUpdateCommand(collectionAndFilter.collection, collectionAndFilter.filter, fieldUpdates));
+    }
+
+    private CollectionAstFilterPair getCollectionAstFilterPair(
+            AbstractUpdateOrDeleteStatement updateOrDeleteStatement) {
+        checkMutationStatement(updateOrDeleteStatement);
+        var fromClause = updateOrDeleteStatement.getFromClause();
+        checkFromClauseSupportability(fromClause);
+        var collection = getMutationCollectionAfterRegisteredAsAffectedTable(updateOrDeleteStatement);
+        var filter = acceptAndYield(updateOrDeleteStatement.getRestriction(), FILTER);
+        return new CollectionAstFilterPair(collection, filter);
     }
 
     @Override
-    public void visitInsertStatement(InsertSelectStatement insertSelectStatement) {
-        throw new FeatureNotSupportedException("TODO-HIBERNATE-46 https://jira.mongodb.org/browse/HIBERNATE-46");
+    public void visitInsertStatement(InsertSelectStatement insertStatement) {
+        checkMutationStatement(insertStatement);
+        if (insertStatement.getConflictClause() != null) {
+            throw new FeatureNotSupportedException();
+        }
+        if (insertStatement.getSourceSelectStatement() != null) {
+            throw new FeatureNotSupportedException();
+        }
+
+        var collection = getMutationCollectionAfterRegisteredAsAffectedTable(insertStatement);
+        var fieldReferences = insertStatement.getTargetColumns();
+        assertFalse(fieldReferences.isEmpty());
+
+        var fieldNames = new ArrayList<String>(fieldReferences.size());
+        for (var fieldReference : fieldReferences) {
+            fieldNames.add(fieldReference.getColumnExpression());
+        }
+
+        var valuesList = insertStatement.getValuesList();
+        assertFalse(valuesList.isEmpty());
+
+        var documents = new ArrayList<AstDocument>(valuesList.size());
+        for (Values values : valuesList) {
+            assertTrue(fieldNames.size() == values.getExpressions().size());
+            var astElements = new ArrayList<AstElement>(values.getExpressions().size());
+            for (var i = 0; i < fieldReferences.size(); i++) {
+                var fieldName = fieldNames.get(i);
+                var fieldValueExpression = values.getExpressions().get(i);
+                if (!isValueExpression(fieldValueExpression)) {
+                    throw new FeatureNotSupportedException();
+                }
+                var fieldValue = acceptAndYield(fieldValueExpression, FIELD_VALUE);
+                astElements.add(new AstElement(fieldName, fieldValue));
+            }
+            documents.add(new AstDocument(astElements));
+        }
+
+        astVisitorValueHolder.yield(COLLECTION_MUTATION, new AstInsertCommand(collection, documents));
```

Comment:
there are 3 assertion usages above as follows:

- insertion field list is not empty
- insertion values list is not empty
- field list length is the same as values length

The above constraints were either enforced from HQL grammar level or semantic checking phase in Hibernate (a prior step when the final SQL AST translation step is reached), so we can rest assured these constraints are enforced already.

Reply:
similarly, various assertions are used in updating case as well. Common constraints have been carefully centralized into `checkMutationStatementSupportability()` method. Only one checking is required to ensure the insertion field value or udpating field value expressions should be either `valueExpression` (literal value or parameter expressions).